Express Response injected is NULL

# 🐞 Bug report

#### What modules are related to this issue?
<!-- ✍️edit: -->
- [ ] aspnetcore-engine
- [ ] builders
- [ ] common
- [x] express-engine
- [ ] hapi-engine
- [ ] module-map-ngfactory-loader

### Is this a regression?

No

### Description

The express response is NULL.

## 🔬 Minimal Reproduction

```ts
import { Inject, PLATFORM_ID, Optional, Injectable } from '@angular/core';
import { REQUEST, RESPONSE } from '@nguniversal/express-engine/tokens';
import { isPlatformServer } from '@angular/common';
import { Request, Response } from 'express';

import Cookies from 'js-cookie'

@Injectable({
  providedIn: 'root'
})
export class CookieService {

  constructor(
    @Inject(PLATFORM_ID) private platformId: any,
    @Optional() @Inject(REQUEST) private request: Request,
    @Optional() @Inject(RESPONSE) private response: Response,
  ) {
  }

  get(key: string) {
    if (isPlatformServer(this.platformId)) {
      return this.request.cookies[key]
    } else {
      return Cookies.get(key)
    }
  }

  getAll() {
    if (isPlatformServer(this.platformId)) {
      return this.request.cookies
    } else {
      return Cookies.get()
    }
  }

  set(key: string, value: string, options: any = undefined) {
    if (isPlatformServer(this.platformId)) {
// THIS IS NULL
      console.log(this.response)
      if (this.response.cookie) {
        this.response.cookie(key, value)
      }
    } else {
      Cookies.set(key, value)
    }
  }
}
```

## 🔥 Exception or Error
<pre><code>
[MASTER] [ERROR] [PID: 010127]  ERROR TypeError: Cannot read property 'cookie' of null
    at CookieService.set (/home/patrikx3/Projects/backwash-ai/dist/backwash-ai/server/en/main.js:202467:31)
    at AppComponent.setTheme (/home/patrikx3/Projects/backwash-ai/dist/backwash-ai/server/en/main.js:202660:21)
    at AppComponent.ngOnInit (/home/patrikx3/Projects/backwash-ai/dist/backwash-ai/server/en/main.js:202654:14)
    at callHook (/home/patrikx3/Projects/backwash-ai/dist/backwash-ai/server/en/main.js:65482:18)
    at callHooks (/home/patrikx3/Projects/backwash-ai/dist/backwash-ai/server/en/main.js:65446:17)
    at executeInitAndCheckHooks (/home/patrikx3/Projects/backwash-ai/dist/backwash-ai/server/en/main.js:65386:9)
    at refreshView (/home/patrikx3/Projects/backwash-ai/dist/backwash-ai/server/en/main.js:71850:21)
    at renderComponentOrTemplate (/home/patrikx3/Projects/backwash-ai/dist/backwash-ai/server/en/main.js:71966:9)
    at tickRootContext (/home/patrikx3/Projects/backwash-ai/dist/backwash-ai/server/en/main.js:73520:9)
    at detectChangesInRootView (/home/patrikx3/Projects/backwash-ai/dist/backwash-ai/server/en/main.js:73554:5)
    at RootViewRef.detectChanges (/home/patrikx3/Projects/backwash-ai/dist/backwash-ai/server/en/main.js:75234:9)
    at ApplicationRef.tick (/home/patrikx3/Projects/backwash-ai/dist/backwash-ai/server/en/main.js:99919:22)
    at ApplicationRef._loadComponent (/home/patrikx3/Projects/backwash-ai/dist/backwash-ai/server/en/main.js:99969:14)
    at ApplicationRef.bootstrap (/home/patrikx3/Projects/backwash-ai/dist/backwash-ai/server/en/main.js:99895:14)
    at /home/patrikx3/Projects/backwash-ai/dist/backwash-ai/server/en/main.js:99491:25
</code></pre>


## 🌍 Your Environment
<pre><code>

     _                      _                 ____ _     ___
    / \   _ __   __ _ _   _| | __ _ _ __     / ___| |   |_ _|
   / △ \ | '_ \ / _` | | | | |/ _` | '__|   | |   | |    | |
  / ___ \| | | | (_| | |_| | | (_| | |      | |___| |___ | |
 /_/   \_\_| |_|\__, |\__,_|_|\__,_|_|       \____|_____|___|
                |___/
    

Angular CLI: 9.1.6
Node: 12.16.3
OS: linux x64

Angular: 9.1.7
... animations, common, compiler, compiler-cli, core, forms
... localize, platform-browser, platform-browser-dynamic
... platform-server, router
Ivy Workspace: Yes

Package                           Version
-----------------------------------------------------------
@angular-devkit/architect         0.901.6
@angular-devkit/build-angular     0.901.6
@angular-devkit/build-optimizer   0.901.6
@angular-devkit/build-webpack     0.901.6
@angular-devkit/core              9.1.6
@angular-devkit/schematics        9.1.6
@angular/cdk                      9.2.3
@angular/cli                      9.1.6
@angular/material                 9.2.3
@ngtools/webpack                  9.1.6
@nguniversal/builders             9.1.1
@nguniversal/common               9.1.1
@nguniversal/express-engine       9.1.1
@schematics/angular               9.1.6
@schematics/update                0.901.6
rxjs                              6.5.5
typescript                        3.8.3
webpack                           4.42.0
</code></pre>

--- a/modules/express-engine/spec/index.spec.ts
+++ b/modules/express-engine/spec/index.spec.ts
@@ -6,6 +6,7 @@ import { SOME_TOKEN } from '../testing/mock.server.module';
 import {
   MockServerModuleNgFactory,
   RequestServerModuleNgFactory,
+  ResponseServerModuleNgFactory,
   TokenServerModuleNgFactory
 } from '../testing/mock.server.module.ngfactory';
 
@@ -49,6 +50,27 @@ describe('test runner', () => {
         throw err;
       }
       expect(html).toContain('url:http://localhost:4200');
+      done();
+    });
+  });
+
+  it('should be able to inject RESPONSE token', (done) => {
+    const someStatusCode = 400;
+    ngExpressEngine({bootstrap: ResponseServerModuleNgFactory})(null as any as string, {
+      req: {
+        get: () => 'localhost',
+        res : {
+          statusCode: someStatusCode
+        }
+      } as any,
+      // TODO this shouldn't be required
+      bootstrap: ResponseServerModuleNgFactory,
+      document: '<root></root>'
+    }, (err, html) => {
+      if (err) {
+        throw err;
+      }
+      expect(html).toContain(`statusCode:${someStatusCode}`);
       done();
     });
   });

--- a/modules/express-engine/src/main.ts
+++ b/modules/express-engine/src/main.ts
@@ -50,6 +50,7 @@ export function ngExpressEngine(setupOptions: Readonly<NgSetupOptions>) {
       }
 
       const req = options.req;
+      const res = options.res || req.res;
 
       const renderOptions: RenderOptions = Object.assign({}, options);
 
@@ -58,8 +59,7 @@ export function ngExpressEngine(setupOptions: Readonly<NgSetupOptions>) {
       renderOptions.document = options.document || getDocument(filePath);
 
       renderOptions.providers = options.providers || [];
-      renderOptions.providers =
-        renderOptions.providers.concat(getReqResProviders(options.req, options.res));
+      renderOptions.providers = renderOptions.providers.concat(getReqResProviders(req, res));
 
       engine.render(renderOptions)
         .then(html => callback(null, html))

--- a/modules/express-engine/testing/mock.server.module.ts
+++ b/modules/express-engine/testing/mock.server.module.ts
@@ -8,7 +8,7 @@
 import { Component, Inject, InjectionToken, NgModule } from '@angular/core';
 import { BrowserModule } from '@angular/platform-browser';
 import { ServerModule } from '@angular/platform-server';
-import { REQUEST } from '@nguniversal/express-engine/tokens';
+import { REQUEST, RESPONSE } from '@nguniversal/express-engine/tokens';
 
 @Component({selector: 'root', template: 'some template'})
 export class MockComponent {
@@ -33,6 +33,19 @@ export class RequestComponent {
   bootstrap: [RequestComponent],
 })
 export class RequestServerModule {
+}
+
+@Component({selector: 'root', template: `statusCode:{{_res.statusCode}}`})
+export class ResponseComponent {
+  constructor(@Inject(RESPONSE) public readonly _res: any) {}
+}
+
+@NgModule({
+  imports: [BrowserModule.withServerTransition({appId: 'mock'}), ServerModule],
+  declarations: [ResponseComponent],
+  bootstrap: [ResponseComponent],
+})
+export class ResponseServerModule {
 }
 
 export const SOME_TOKEN = new InjectionToken<string>('SOME_TOKEN');

